troposphere/emr.py

```diff
@@ -0,0 +1,186 @@
+# Copyright (c) 2012-2013, Antonio Alonso Dominguez <alonso.domin@gmail.com>
+# All rights reserved.
+#
+# See LICENSE file for full license.
+
+from . import AWSObject, AWSProperty, AWSHelperFn
+from .validators import (boolean, integer)
+
+
+class KeyValue(AWSHelperFn):
+    def __init__(self, key, value):
+        self.data = {'Key': key, 'Value': value}
+
+    def JSONrepr(self):
+        return self.data
+
+
+def additional_info_validator(xs):
+    if not isinstance(xs, dict):
+        raise ValueError("AdditionalInfo must be a dict of "
+                         "string to string pairs")
+    for k, v in xs.iteritems():
+        if not isinstance(k, basestring):
+            raise ValueError('AdditionalInfo keys must be strings')
+        if not isinstance(v, basestring):
+            raise ValueError('AdditionalInfo values must be strings')
+
+    return xs
+
+
+class Application(AWSProperty):
+    props = {
+        'AdditionalInfo': (additional_info_validator, False),
+        'Args': (list, False),
+        'Name': (basestring, False),
+        'Version': (basestring, False)
+    }
+
+
+class ScriptBootstrapActionConfig(AWSProperty):
+    props = {
+        'Args': ([basestring], False),
+        'Path': (basestring, True)
+    }
+
+
+class BootstrapActionConfig(AWSProperty):
+    props = {
+        'Name': (basestring, True),
+        'ScriptBootstrapAction': (ScriptBootstrapActionConfig, True)
+    }
+
+
+def properties_validator(xs):
+    if not isinstance(xs, dict):
+        raise ValueError("ConfigurationProperties must be a dict of "
+                         "string to string pairs")
+    for k, v in xs.iteritems():
+        if not isinstance(k, basestring):
+            raise ValueError('ConfigurationProperties keys must be strings')
+        if not isinstance(v, basestring):
+            raise ValueError('ConfigurationProperties values must be strings')
+
+    return xs
+
+
+def configurations_validator(xs):
+    if not isinstance(xs, list):
+        raise ValueError("Configurations must be a list of "
+                         "Configuration objects.")
+    for x in xs:
+        if not isinstance(x, Configuration):
+            raise ValueError("Configuration '%s' must be of "
+                             "Configuration type" % x)
+    return xs
+
+
+class Configuration(AWSProperty):
+    props = {
+        'Classification': (basestring, False),
+        'ConfigurationProperties': (properties_validator, False),
+        'Configurations': (configurations_validator, False)
```

Comment:
Main reason I chose for this custom validator function in this specific case is because Python won't allow me to _self reference_ the `Configuration` type in the nested `Configurations` property.